projects/packages/forms/src/blocks/shared/hooks/use-synced-attributes.js

```diff
@@ -0,0 +1,187 @@
+ * @param {object} props.children - Child components to render within the provider.
+ * @return {JSX.Element}          The provider component.
+ */
+export function SyncedAttributeProvider( { children } ) {
+	const value = useReducer( ( state, action ) => {
+		const { name, attributes } = action;
+		return {
+			...state,
+			[ name ]: {
+				...( state[ name ] ?? {} ),
+				...attributes,
+			},
+		};
+	}, {} );
+
+	return (
+		<SyncedAttributeContext.Provider value={ value }>{ children }</SyncedAttributeContext.Provider>
+	);
+}
+
+/**
+ * Uses the `SyncedAttributeContext`, but returns only the values for the given block type.
+ * The `setSyncedAttributes` function it returns is also adjusted so that it only updates
+ * the attributes for the given block type.
+ *
+ * @param {string} name - The name of the block.
+ *
+ * @return {Array} Array containing the synced attributes and the function to update them.
+ */
+function useSyncedAttributesForBlock( name ) {
+	const [ syncedAttributes, setSyncedAttributes ] = useContext( SyncedAttributeContext );
+
+	return useMemo( () => {
+		return [
+			syncedAttributes?.[ name ],
+			attributes => setSyncedAttributes( { name, attributes } ),
+		];
+	}, [ name, setSyncedAttributes, syncedAttributes ] );
+}
+
+/**
+ * Returns an object containing only the attributes that are in the syncedAttributeKeys array.
+ *
+ * @param {object}   attributes          - The attributes of the block.
+ * @param {string[]} syncedAttributeKeys - The keys of the attributes that are synced.
+ *
+ * @return {object} Object containing only the synced attributes.
+ */
+function pickSyncedAttributes( attributes, syncedAttributeKeys ) {
+	return syncedAttributeKeys.reduce( ( acc, key ) => {
+		acc[ key ] = attributes[ key ];
+		return acc;
+	}, {} );
+}
+
+/**
+ * Returns an object containing only the attributes that have different values compared to the attribute sets.
+ *
+ * When multiple attribute sets are provided, the function performs an AND check, so for a given key,
+ * the value in EVERY attribute set must be different for it to be considered modified.
+ *
+ * @param {object}   sourceAttributes - The source attributes to compare against.
+ * @param {Object[]} attributeSets    - Array of attribute sets to compare with the source.
+ *
+ * @return {object} Object containing only the modified attributes.
+ */
+function getModifiedAttributes( sourceAttributes, attributeSets ) {
+	if ( ! sourceAttributes ) {
+		return {};
+	}
+
+	return Object.keys( sourceAttributes ).reduce( ( acc, key ) => {
+		// Use lodash `isEqual` to perform a deep comparison of the attributes.
+		// With `!==` this would return a lot of false positives where different blocks
+		// have objects with different references, but the same inner attribute values.
+		// TODO - replace this with not lodash.
+		if ( attributeSets.every( set => ! isEqual( sourceAttributes[ key ], set?.[ key ] ) ) ) {
+			acc[ key ] = sourceAttributes[ key ];
+		}
```

Comment:
I believe core generally uses `fast-deep-equal` over Lodash's `isEqual`. Can we use that here?